experimental/lib/Transforms/Speculation/PlacementFinder.cpp

```diff
@@ -0,0 +1,359 @@
+        placements.addCommit(res, succOp);
+      } else if (isa<handshake::EndOp>(succOp)) {
+        // A commit is needed in front of the end/exit operation
+        placements.addCommit(res, succOp);
+      } else {
+        findCommitsTraversal(visited, succOp);
+      }
+    }
+  }
+}
+
+struct endpointComparator {
+  bool operator()(const BBEndpoints &a, const BBEndpoints &b) const {
+    if (a.srcBB != b.srcBB)
+      return a.srcBB < b.srcBB;
+    return a.dstBB < b.dstBB;
+  }
+};
+using BBEndpointsMap =
+    std::map<BBEndpoints, std::vector<OpPlacement>, endpointComparator>;
+
+// Data structure to hold all arcs leading to a single BB predecessor
+// Note: srcBB and dstBB can be equal when the arcs are Backedges
+struct BlockPredecessor {
+  unsigned srcBB;
+  unsigned dstBB;
+  std::vector<OpPlacement> arcs;
+};
+using BBPredecessorMap = std::map<unsigned, std::vector<BlockPredecessor>>;
+
+// Get a map from BBs to the list of its BlockPredecessors
+static BBPredecessorMap getBlockPredecessors(handshake::FuncOp funcOp) {
+  // Traverse all operations to find arcs between BBs (and self-arcs)
+  BBEndpointsMap endpointArcs;
+  funcOp->walk([&](Operation *op) {
+    for (Value operand : op->getOperands()) {
+      BBEndpoints endpoints;
+      if (isBackedge(operand, op, &endpoints) or
+          endpoints.srcBB != endpoints.dstBB) {
+        OpPlacement arc = {operand, op};
+        endpointArcs[endpoints].push_back(arc);
+      }
+    }
+  });
+
+  // Join all predecessors of a BB
+  BBPredecessorMap predecessors;
+  for (const auto &[endpoints, arcs] : endpointArcs) {
+    BlockPredecessor predecessor;
+    predecessor.srcBB = endpoints.srcBB;
+    predecessor.dstBB = endpoints.dstBB;
+    predecessor.arcs = arcs;
+    predecessors[endpoints.dstBB].push_back(predecessor);
+  }
+
+  return predecessors;
+}
+
+// DFS traversal to mark all operations that lead to Commit units
+static void markSpeculativePaths(Operation *currOp,
+                                 SpeculationPlacements &placements,
+                                 PlacementList &markedPaths) {
+  for (Value res : currOp->getResults()) {
+    for (Operation *succOp : res.getUsers()) {
+      OpPlacement arc = {res, succOp};
+      if (not markedPaths.count(arc)) {
+        markedPaths.insert(arc);
+        // Stop traversal if a commit is reached
+        if (not placements.containsCommit(res, succOp))
+          markSpeculativePaths(succOp, placements, markedPaths);
+      }
+    }
+  }
+}
+
+void PlacementFinder::findCommitsBetweenBBs() {
+  OpPlacement specPos = placements.getSpeculatorPlacement();
+  auto funcOp = specPos.dstOp->getParentOfType<handshake::FuncOp>();
+  assert(funcOp && "op should have parent function");
+
+  // Whenever a BB has two speculative inputs, commit units are needed to avoid
+  // tokens going out-of-order. First, the block predecessor arcs are found
+  BBPredecessorMap predecessors = getBlockPredecessors(funcOp);
```

Comment:
this also should not be called predecessors. I had no idea what you were iterating over, but now I see you're just iterating over all the BBs. I normally name these things something like bbToArcsMap

Reply:
The naming of this whole section has been improved to better describe what the objects do.